### PR TITLE
[Modelzoo] Create MMoE model.

### PR DESCRIPTION
**Pull request by jmaksymc**
This PR is with distribute training implementation since I have no acces to @kmagiers MMoE fork.
Jira [WC1-1042]

### CODE REVIEW

File: modelzoo/MMoE/train.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+        self.__feature_column = feature_column
+        self.__num_experts = num_experts
+        self.__expert_hidden_units = expert_hidden_units
+        self.__towers = towers
+        self.__l2_regularization = l2_regularizer(l2_scale) if l2_scale else None
+        self.__learning_rate = learning_rate
+        self.__bf16 = bf16
+        self.__input_layer_partitioner = input_layer_partitioner
+        self.__dense_layer_partitioner = dense_layer_partitioner
+        self.__is_training = is_training
```
Comment (chenzhiyin) [inline modelzoo/MMoE/train.py:199]: the variable "self.__is_training" is not used in the code.

File: modelzoo/MMoE/train.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+                                                'MMOE_checkpoint'),
+                                                global_step=_in)
+                    print(f'Save checkpoint to {checkpoint_path}')
+                elif (args.timeline > 0 and _in % args.timeline == 0):
+                    _, train_loss = sess.run([model.train_op, model.loss],
+                                            options=options,
+                                            run_metadata=run_metadata)
+                    fetched_timeline = timeline.Timeline(run_metadata.step_stats)
+                    chrome_trace = fetched_timeline.generate_chrome_trace_format()
+                    print(f'Save timeline to {checkpoint_path}')
```
Comment (chenzhiyin) [inline modelzoo/MMoE/train.py:390]: “checkpoint_path” should be "checkpoint_drive".  （This comment is proposed by Wang Duyi）

File: modelzoo/MMoE/train.py [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+        for i in range(0, len(LABEL_COLUMNS)):
+            labels.append(all_columns.pop(LABEL_COLUMNS[i]))
+        label = tf.stack(labels, axis=1)
+        features = all_columns
+        return features, label
+
+    # Extract lines from input files using the Dataset API.
+    dataset = tf.data.TextLineDataset(filename)
+    dataset = dataset.shuffle(buffer_size=400000,
+                              seed=SEED)  # set seed for reproducing
```
Comment (chenzhiyin) [inline modelzoo/MMoE/train.py:95]: the argument "seed" is not effective here.

File: modelzoo/MMoE/train.py [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+                master=server.target,
+                is_chief=tf_config['is_chief'],
+                checkpoint_dir=checkpoint_dir,
+                scaffold=scaffold,
+                hooks=hooks,
+                log_step_count_steps=100,
+                config=sess_config) as sess:
+            while not sess.should_stop():
+                _, train_loss = sess.run([model.train_op, model.loss])
+    else:
```
Comment (chenzhiyin) [inline modelzoo/MMoE/train.py:358]: In the new template, we only support “MonitoredTrainingSession”.

File: modelzoo/MMoE/train.py (hunk 1 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Removed self.__is_training.

File: modelzoo/MMoE/train.py (hunk 2 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Changed `checkpoint_path` to `checkpoint_dir`

File: modelzoo/MMoE/train.py (hunk 3 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Changed.

File: modelzoo/MMoE/train.py [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+                
+                train(model,
+                      args.output_dir,
+                      train_dataset,
+                      test_dataset,
+                      args.keep_checkpoint_max,
+                      train_steps,
+                      test_steps,
+                      no_eval=args.no_eval,
+                      timeline_steps=args.timeline,
```
[the remark this answers is not in the document]
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Removed.

File: modelzoo/MMoE/train.py [hunk 6]
```diff
@@ -0,0 +1,673 @@
+from tabnanny import check
+import time
+import argparse
+import numbers
+from unittest.mock import NonCallableMagicMock
```
[the remark this answers is not in the document]
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Removed unnecessary imports.

File: modelzoo/MMoE/train.py [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+                                                            name=f'{expert_layer_scope.name}/dense')
+                            expert_features = DNN_ACTIVATION(expert_features, 
+                                                            name=f'{expert_layer_scope.name}/act')               
+                            add_layer_summary(expert_features, expert_layer_scope.name)
+                    experts.append(expert_features)
+            experts_features = tf.stack(experts, axis=1)
+
+            towers=[]
+            for [tower_name, label_name, hidden_units] in self.__towers:
+                gate_input = input_emb
```
[the remark this answers is not in the document]
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Changed

**Comment by changqi1** [comment]
Test AUC, ACC, Gsteps/sec on CPX

|  | ACC | |
| -- | -- | -- |
|   | value | percent |
| Community TF | 0.973787 | baseline |
| DeepRec FP32 | 0.973787 | 100.00% |
| DeepRec BF16 | 0.973787 | 100.00% |

|  | AUC | |
| -- | -- | --  |
|   | value | percent  |
| Community TF | 0.7432368 | baseline |
| DeepRec FP32 | 0.7444535 | 100.16% |
| DeepRec BF16 | 0.7453669 | 100.29% |

|  | Gsteps/s | |
| -- | -- | -- |
|   | value | percent |
| Community TF | 73.74488 | baseline |
| DeepRec FP32 | 93.01421 | 126.13% |
| DeepRec BF16 | 105.70906 | 143.34% |
